apps/i18n.py

```diff
@@ -0,0 +1,178 @@
+    data = {}
+    messages = Set()
+    universal_messages = Set()
+    for path in files:
+        locale = locale_from_filename(path)
+        if locale not in data:
+            data[locale] = {}
+        with open(path, "r") as file:
+            for line in file:
+                name,definition = split_line(line)
+                if locale == "universal":
+                    if name in messages:
+                        sys.stderr.write("Error: Redefinition of message \"" + name + "\" as universal\n")
+                        sys.exit(-1)
+                    if name in universal_messages:
+                        sys.stderr.write("Error: Redefinition of universal message \"" + name + "\"\n")
+                        sys.exit(-1)
+                    universal_messages.add(name)
+                else:
+                    messages.add(name)
+                data[locale][name] = definition
+    return {"messages": sorted(messages), "universal_messages": sorted(universal_messages), "data": data}
+
+def print_header(data, path, locales):
+    f = open(path, 'w')
+    f.write("#ifndef APPS_I18N_H\n")
+    f.write("#define APPS_I18N_H\n")
+    f.write("\n")
+    f.write("// This file is auto-generated by i18n.py\n")
+    f.write("\n")
+    f.write("#include <escher.h>\n")
+    f.write("\n")
+    f.write("namespace I18n {\n")
+    f.write("\n")
+    f.write("constexpr static int NumberOfLanguages = %d;\n" % len(locales))
+    f.write("\n")
+    f.write("enum class Message : uint16_t {\n")
+    for message in data["messages"]:
+        f.write("  " + message + ",\n")
+    f.write("\n")
+    f.write("  UniversalMessageMarker,\n")
+    f.write("\n")
+    for message in data["universal_messages"]:
+        f.write("  " + message + ",\n")
+    f.write("};\n")
+    f.write("\n")
+    f.write("enum class Language : uint16_t {\n")
+    f.write("  Default = 0,\n")
+    for locale in locales:
+        f.write("  " + locale.upper() + ",\n")
+    f.write("};\n")
+    f.write("\n")
+    f.write("constexpr const Message LanguageNames[NumberOfLanguages] = {\n");
+    for locale in locales:
+        f.write("  Message::Language" + locale.upper() + ",\n")
+    f.write("};\n")
+    f.write("\n")
+    f.write("}\n")
+    f.write("\n")
+    f.write("#endif\n")
+    f.close()
+
+def print_implementation(data, path, locales):
+    f = open(path, 'w')
+    f.write("#include \"i18n.h\"\n")
+    f.write("#include \"global_preferences.h\"\n")
+    f.write("#include <assert.h>\n");
+    f.write("\n")
+    f.write("namespace I18n {\n")
+    f.write("\n")
+    for message in data["messages"]:
+        for locale in locales:
+            if not locale in data["data"]:
+                sys.stderr.write("Error: Undefined locale \"" + locale + "\"\n")
+                sys.exit(-1)
+            if not message in data["data"][locale]:
+                sys.stderr.write("Error: Undefined key \"" + message + "\" for locale \"" + locale + "\"\n")
+                sys.exit(-1)
```

Comment:
That's a bit extreme for a missing translation. Third-party contributors/app developers like me typically aren't fluent in all of Dutch, English, Spanish, French and Portuguese (and probably more in the future).

If an English translation of a missing key is available, I'd output a warning and use it. At least Travis won't fail our pull requests just because we aren't polyglots and we can actually see what's truly missing, instead of blindly padding with junk to keep the compiler happy (like in #163).

Reply:
At the moment we want to support all 5 built-in languages (en,es,fr,de,pt) so we wouldn't merge a feature without the appropriate translation anyway.

That being said, you make a very good point about contributors not being fluent in all those languages. That's not a big issue though: we'll help :)